library/lua/gui.lua

```diff
@@ -178,8 +208,31 @@ end
 -- Clipped view rectangle object --
 -----------------------------------
 
+---@alias gui.dimension { x1: integer, y1: integer, x2: integer, y2: integer, width: integer, height: integer }
+
+---@class gui.ViewRectAttrs
+---@field rect? gui.dimension
+---@field clip_rect? gui.dimension
+---@field view_rect? gui.ViewRect
+---@field clip_view? gui.ViewRect
+---@field clip_x1? integer
+---@field clip_y1? integer
+---@field clip_x2? integer
+---@field clip_y2? integer
+---@field x1? integer
+---@field y1? integer
+---@field x2? integer
+---@field y2? integer
+---@field width? integer
+---@field height? integer
+
+---@class gui.ViewRect: dfhack.class, gui.ViewRectAttrs
+---@field ATTRS fun(attributes: gui.ViewRectAttrs)
+---@overload fun(attributes: gui.ViewRectAttrs): self
 ViewRect = defclass(ViewRect, nil)
 
+---@param self gui.ViewRect
+---@param args gui.ViewRectAttrs
```

Comment:
I wrote up a bit response and I think Github lost it. Essentially the issue here is that since lua language server lacks a generic type `Partial<T>` we are stuck with a choice of making the base class entirely optional or having to write both an instance version of the type and an init version of the type, both having mostly the same fields...